consensus/src/sender_aware_shuffler.rs

```diff
@@ -0,0 +1,466 @@
+/// invariant in terms of ordering
+/// 1. Relative ordering of all transactions from the same before and after shuffling is same
+/// 2. Relative ordering of all transactions across different senders will also be maintained if they are
+/// non-conflicting. In other words, if the input block has only one transaction per sender, the output
+/// ordering will remain unchanged.
+///
+/// The shuffling algorithm is O(n) and following is the pseudo code for it.
+/// loop:
+///   if a sender fell out of the sliding window in previous iteration,
+///      then: we add the first pending transaction from that sender to the block
+///   else while we have transactions to process in the original transaction order
+///         take a new one,
+///         if it conflicts, add to the pending set
+///         else we add it to the block
+///   else
+///       take the first transaction from the pending transactions and add it to the block
+
+pub struct SenderAwareShuffler {
+    conflict_window_size: usize,
+}
+
+impl TransactionShuffler for SenderAwareShuffler {
+    fn shuffle(&self, txns: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let _timer = TXN_SHUFFLE_SECONDS.start_timer();
+
+        // maintains the intermediate state of the shuffled transactions
+        let mut sliding_window = SlidingWindowState::new(self.conflict_window_size, txns.len());
+        let mut pending_txns = PendingTransactions::new();
+        let num_transactions = txns.len();
+        let mut orig_txns = VecDeque::from(txns);
+        let mut next_to_add =
+            |sliding_window: &mut SlidingWindowState| -> Option<SignedTransaction> {
+                if sliding_window.num_txns() == num_transactions {
+                    return None;
+                }
+                // First check if we have a sender dropped off of conflict window in previous step, if so,
+                // we try to find pending transaction from the corresponding sender and add it to the block.
+                if let Some(sender) = sliding_window.last_dropped_sender() {
+                    if let Some(txn) = pending_txns.remove_pending_from_sender(sender) {
+                        return Some(txn);
+                    }
+                }
+                // If we can't find any transaction from a sender dropped off of conflict window, then
+                // iterate through the original transactions and try to find the next candidate
+                while let Some(txn) = orig_txns.pop_front() {
+                    if !sliding_window.has_conflict(&txn.sender()) {
+                        return Some(txn);
+                    }
+                    pending_txns.add_transaction(txn);
+                }
+
+                // If we can't find any candidate in above steps, then lastly
+                // add pending transactions in the order if we can't find any other candidate
+                Some(pending_txns.remove_first_pending().unwrap())
```

Comment:
oh I guess I misread the condition here, seems it's simpler to just return Txn instead of Option here and do 
```
while sliding_window.len() < num_transactions {
  sliding_window_add_transaction(next_to_add(&mut sliding_window));
}
```